README.md
## adhocore/cron-expr

[![Latest Version](https://img.shields.io/github/release/adhocore/php-cron-expr.svg?style=flat-square)](https://github.com/adhocore/php-cron-expr/releases)
[![Travis Build](https://img.shields.io/travis/adhocore/php-cron-expr/master.svg?style=flat-square)](https://travis-ci.org/adhocore/php-cron-expr?branch=master)
[![Scrutinizer CI](https://img.shields.io/scrutinizer/g/adhocore/php-cron-expr.svg?style=flat-square)](https://scrutinizer-ci.com/g/adhocore/php-cron-expr/?branch=master)
[![Codecov branch](https://img.shields.io/codecov/c/github/adhocore/php-cron-expr/master.svg?style=flat-square)](https://codecov.io/gh/adhocore/php-cron-expr)
[![StyleCI](https://styleci.io/repos/94228363/shield)](https://styleci.io/repos/94228363)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Donate 15](https://img.shields.io/badge/donate-paypal-blue.svg?style=flat-square&label=donate+15)](https://www.paypal.me/ji10/15usd)
[![Donate 25](https://img.shields.io/badge/donate-paypal-blue.svg?style=flat-square&label=donate+25)](https://www.paypal.me/ji10/25usd)
[![Donate 50](https://img.shields.io/badge/donate-paypal-blue.svg?style=flat-square&label=donate+50)](https://www.paypal.me/ji10/50usd)
[![Tweet](https://img.shields.io/twitter/url/http/shields.io.svg?style=social)](https://twitter.com/intent/tweet?text=Lightweight+and+fast+cron+expression+parser+for+PHP&url=https://github.com/adhocore/php-cron-expr&hashtags=php,cron,cronparser,parser,cronexpr)


- Lightweight Cron expression parser library for PHP.
- Zero dependency.
- Very **fast** because it bails early in case a segment doesnt match.
- Real [benchmark](https://github.com/adhocore/php-cron-bench) shows it is about 7.54x to 12.92x faster than `dragonmantank/cron-expression`

## Installation

```bash
composer require adhocore/cron-expr

# PHP5.6 or lower
composer require adhocore/cron-expr:0.1.0
```

## Usage

**Basic**

```php
use Ahc\Cron\Expression;
use Ahc\Cron\Normalizer;

Expression::isDue('@always');
Expression::isDue(Normalizer::HOURLY, '2015-01-01 00:00:00');
Expression::isDue('*/20 * * * *', new DateTime);
Expression::isDue('5-34/4 * * * *', time());

// Dont like static calls? Below is possible too!
$expr = new Expression;
$expr->isCronDue('*/1 * * * *', time());
```

**Bulk checks**

When checking for several jobs at once, if more than one of the jobs share equivalent expression
then the evaluation is done only once per go thus greatly improving performnce.

```php
use Ahc\Cron\Expression;

$jobs = [
    'job1' => '*/2 */2 * * *',
    'job1' => '* 20,21,22 * * *',
    'job3' => '7-9 * */9 * *',
    'job4' => '*/5 * * * *',
    'job5' => '@5minutes',     // equivalent to job4 (so it is due if job4 is due)
    'job6' => '7-9 * */9 * *', // exact same as job3 (so it is due if job3 is due)
];

// The second param $time can be used same as above: null/time()/date string/DateTime
$dues = Expression::getDues($jobs, '2015-08-10 21:50:00');
// ['job1', 'job4', 'job5']

// Dont like static calls? Below is possible too!
$expr = new Expression;
$dues = $expr->filter($jobs, time());
```

### Cron Expression

Cron expression normally consists of 5 segments viz:
```
<minute> <hour> <day> <month> <weekday>
```
and sometimes there can be 6th segment for `<year>` at the end.

### Real Abbreviations

You can use real abbreviations for month and week days. eg: `JAN`, `dec`, `fri`, `SUN`

### Tags

Following tags are available and they are converted to real cron expressions before parsing:

- *@yearly* or *@annually* - every year
- *@monthly* - every month
- *@daily* - every day
- *@weekly* - every week
- *@hourly* - every hour
- *@5minutes* - every 5 minutes
- *@10minutes* - every 10 minutes
- *@15minutes* - every 15 minutes
- *@30minutes* - every 30 minutes
- *@always* - every minute

> You can refer them with constants from `Ahc\Cron\Normalizer` like `Ahc\Cron\Normalizer::WEEKLY`

### Modifiers

Following modifiers supported

- *Day of Month / 3rd segment:*
    - `L` stands for last day of month (eg: `L` could mean 29th for February in leap year)
    - `W` stands for closest week day (eg: `10W` is closest week days (MON-FRI) to 10th date)
- *Day of Week / 5th segment:*
    - `L` stands for last weekday of month (eg: `2L` is last monday)
    - `#` stands for nth day of week in the month (eg: `1#2` is second sunday)

## LICENSE

> &copy; [MIT](./LICENSE) | 2017-2019, Jitendra Adhikari

## Credits

This project is release managed by [please](https://github.com/adhocore/please).


CHANGELOG.md
## [1.1.3](https://github.com/adhocore/php-cron-expr/releases/tag/1.1.3) (2021-10-20)

### Bug Fixes
- **Validator**: Strict checks for in-step, handle weekends correctly (Jitendra Adhikari) [_b894e4e_](https://github.com/adhocore/php-cron-expr/commit/b894e4e)
- **Normalizer**: Expr maybe lowercased (Jitendra Adhikari) [_e6c11df_](https://github.com/adhocore/php-cron-expr/commit/e6c11df)


## [1.1.2](https://github.com/adhocore/php-cron-expr/releases/tag/1.1.2) (2021-02-20)

### Features
- Add expression normalizer (Jitendra Adhikari) [_afb57c4_](https://github.com/adhocore/php-cron-expr/commit/afb57c4)

### Internal Refactors
- Use Normalizer instead (Jitendra Adhikari) [_42700ab_](https://github.com/adhocore/php-cron-expr/commit/42700ab)

### Documentations
- Tags as constants (Jitendra Adhikari) [_1382d7d_](https://github.com/adhocore/php-cron-expr/commit/1382d7d)


## [1.1.1](https://github.com/adhocore/php-cron-expr/releases/tag/1.1.1) (2020-01-09)

### Miscellaneous
- **Validator**: Cleanup redundancy (Jitendra Adhikari) [_00983d4_](https://github.com/adhocore/php-cron-expr/commit/00983d4)
- **Travis**: Script (Jitendra Adhikari) [_fec5332_](https://github.com/adhocore/php-cron-expr/commit/fec5332)
- **Composer**: Tweak script.test (Jitendra Adhikari) [_0a5b4fb_](https://github.com/adhocore/php-cron-expr/commit/0a5b4fb)

### Documentations
- Update benchmark (Jitendra Adhikari) [_fe9dffd_](https://github.com/adhocore/php-cron-expr/commit/fe9dffd)


## [1.1.0](https://github.com/adhocore/php-cron-expr/releases/tag/1.1.0) (2019-12-27)

### Features
- Add ref time class (Jitendra Adhikari) [_2ad504b_](https://github.com/adhocore/php-cron-expr/commit/2ad504b)

### Bug Fixes
- **Expr**: Replace literals case insensitive (Jitendra Adhikari) [_a5c179f_](https://github.com/adhocore/php-cron-expr/commit/a5c179f)

### Internal Refactors
- **Validator**: Use reference time class (Jitendra Adhikari) [_05f139d_](https://github.com/adhocore/php-cron-expr/commit/05f139d)
- **Checker**: Use reference time class (Jitendra Adhikari) [_7b4138f_](https://github.com/adhocore/php-cron-expr/commit/7b4138f)
- **Expr**: Use reference time class, cleanup process() (Jitendra Adhikari) [_1ce873d_](https://github.com/adhocore/php-cron-expr/commit/1ce873d)

### Miscellaneous
- **Reftime**: Add method annot for magic calls (Jitendra Adhikari) [_96b78d1_](https://github.com/adhocore/php-cron-expr/commit/96b78d1)
- Ignore coverage xml (Jitendra Adhikari) [_2a9505a_](https://github.com/adhocore/php-cron-expr/commit/2a9505a)
- **Composer**: Bump deps version, fix test:cov (Jitendra Adhikari) [_14e9117_](https://github.com/adhocore/php-cron-expr/commit/14e9117)

### Documentations
- About cron (Jitendra Adhikari) [_a3760f8_](https://github.com/adhocore/php-cron-expr/commit/a3760f8)


## [1.0.0](https://github.com/adhocore/php-cron-expr/releases/tag/1.0.0) (2019-12-22)

### Internal Refactors
- Strict php7 typehints (Jitendra Adhikari) [_e0967be_](https://github.com/adhocore/php-cron-expr/commit/e0967be)

### Miscellaneous
- Add composer script (Jitendra Adhikari) [_594e7e1_](https://github.com/adhocore/php-cron-expr/commit/594e7e1)

### Documentations
- Update for v1.0 (Jitendra Adhikari) [_7224037_](https://github.com/adhocore/php-cron-expr/commit/7224037)

### Builds
- **Travis**: Php7 only (Jitendra Adhikari) [_b0170db_](https://github.com/adhocore/php-cron-expr/commit/b0170db)


## [v0.1.0](https://github.com/adhocore/php-cron-expr/releases/tag/v0.1.0) (2019-09-22)

### Documentations
- About stability (Jitendra Adhikari) [_1672edc_](https://github.com/adhocore/php-cron-expr/commit/1672edc)
- Add php support info (Jitendra Adhikari) [_9d21717_](https://github.com/adhocore/php-cron-expr/commit/9d21717)


## [v0.0.7](https://github.com/adhocore/php-cron-expr/releases/tag/v0.0.7) (2019-08-12)

### Internal Refactors
- **Expr**: Normalize expr, use regex split instead (Jitendra Adhikari) [_74f8dfc_](https://github.com/adhocore/php-cron-expr/commit/74f8dfc)


## [v0.0.6] 2018-08-16 00:08:45 UTC

- [d933099](https://github.com/adhocore/php-cron-expr/commit/d933099) fix(expr): static ::instance()
- [48eef4a](https://github.com/adhocore/php-cron-expr/commit/48eef4a) docs: bulk checks/filters
- [8c23489](https://github.com/adhocore/php-cron-expr/commit/8c23489) test(expr): ...
- [aab941e](https://github.com/adhocore/php-cron-expr/commit/aab941e) refactor(expr): simplify filter(), also cache undue case
- [cd3cd33](https://github.com/adhocore/php-cron-expr/commit/cd3cd33) test(expr): filter(), getDues()
- [7ec07c8](https://github.com/adhocore/php-cron-expr/commit/7ec07c8) feat(checker): make validator injectable
- [0d6d85a](https://github.com/adhocore/php-cron-expr/commit/0d6d85a) feat(expr): add construct with injection/initiliazation
- [b8aa4ce](https://github.com/adhocore/php-cron-expr/commit/b8aa4ce) feat(expr): add ::instance() api
- [cee9bf7](https://github.com/adhocore/php-cron-expr/commit/cee9bf7) feat(expr): add filter(), ::getDues() and normalizeExpr()
- [d73a078](https://github.com/adhocore/php-cron-expr/commit/d73a078) refactor(docblocks): document public apis/methods
- [ee2cc96](https://github.com/adhocore/php-cron-expr/commit/ee2cc96) refactor: rename array of time parts to times

## [v0.0.5] 2018-08-15 10:08:16 UTC

- [6fccf54](https://github.com/adhocore/php-cron-expr/commit/6fccf54) test(validator): instep changes
- [93d2f7f](https://github.com/adhocore/php-cron-expr/commit/93d2f7f) fix(validator): inStep when start+step>end | closes #12
- [895eade](https://github.com/adhocore/php-cron-expr/commit/895eade) docs: update readme
- [bc65912](https://github.com/adhocore/php-cron-expr/commit/bc65912) docs: fix badge


tests/ExpressionTest.php
<?php

/*
 * This file is part of the PHP-CRON-EXPR package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\Cron\Test;

use Ahc\Cron\Expression;
use PHPUnit\Framework\TestCase;

class ExpressionTest extends TestCase
{
    /**
     * @dataProvider scheduleProvider
     */
    public function test_isDue($expr, $time, $foo, $expected, $throwsAt = false)
    {
        $actual = Expression::isDue($expr, $time);

        $this->assertSame($expected, $actual, 'The expression ' . $expr . ' has failed');
    }

    /**
     * @dataProvider invalidScheduleProvider
     * @expectedException \UnexpectedValueException
     */
    public function test_isDue_on_invalid_expression($expr, $time, $foo, $expected, $throwsAt = false)
    {
        Expression::isDue($expr, $time);
    }

    public function test_isCronDue()
    {
        $expr = new Expression;

        $this->assertInternalType('boolean', $expr->isCronDue('*/1 * * * *', time()));
    }

    /**
     * @expectedException \UnexpectedValueException
     */
    public function test_isDue_throws_if_expr_invalid()
    {
        Expression::isDue('@invalid');
    }

    /**
     * @expectedException \UnexpectedValueException
     */
    public function test_isDue_throws_if_modifier_invalid()
    {
        Expression::isDue('* * 2L * *');
    }

    public function test_filter_getDues()
    {
        $jobs = [
            'job1' => '*/2 */2 * * *',
            'job1' => '* 20,21,22 * * *',
            'job3' => '7-9 * */9 * *',
            'job4' => '*/5 * * * *',
            'job5' => '@5minutes',
            'job6' => '7-9 * */9 * *',
        ];

        $this->assertSame(['job1', 'job4', 'job5'], Expression::getDues($jobs, '2015-08-10 21:50:00'));
    }

    /**
     * Data provider for cron schedule.
     *
     * @return array
     */
    public function invalidScheduleProvider()
    {
        return [
            ['* * * * 4W', strtotime('2011-07-01 00:00:00'), '2011-07-27 00:00:00', false, 4], // seg 4
            ['* * * 1L *', strtotime('2011-07-01 00:00:00'), '2011-07-27 00:00:00', false, 3], // seg 3
        ];
    }

    /**
     * The test cases are taken from awesome mtdowling/cron-expression package. Thank you.
     *
     * @link https://github.com/mtdowling/cron-expression/
     *
     * Data provider for cron schedule
     *
     * @return array
     */
    public function scheduleProvider()
    {
        return [
            ['@always', time(), '', true],
            ['* * * * * 2018', null, '', false],
            ['* * * * * 2018', new \DateTime, '', false],
            ['@5minutes', new \DateTime('2017-05-10 02:30:00'), '', true],
            ['* * 7W * *', '2017-10-15 20:00:00', '', false],

            ['*/2 */2 * * *', '2015-08-10 21:47:27', '2015-08-10 22:00:00', false],
            ['* * * * *', '2015-08-10 21:50:37', '2015-08-10 21:50:00', true],
            ['* * * * * ', '2015-08-10 21:50:37', '2015-08-10 21:50:00', true],
            ['*  *  *  *  *', '2015-08-10 21:50:37', '2015-08-10 21:50:00', true],
            ["*\t*\t*\t*\t*", '2015-08-10 21:50:37', '2015-08-10 21:50:00', true],
            ["*\t\t*\t\t*\t\t*\t\t*", '2015-08-10 21:50:37', '2015-08-10 21:50:00', true],
            ['* 20,21,22 * * *', '2015-08-10 21:50:00', '2015-08-10 21:50:00', true],
            // Handles CSV values
            ['* 20,22 * * *', '2015-08-10 21:50:00', '2015-08-10 22:00:00', false],
            // CSV values can be complex
            ['* 5,21-22 * * *', '2015-08-10 21:50:00', '2015-08-10 21:50:00', true],
            ['7-9 * */9 * *', '2015-08-10 22:02:33', '2015-08-18 00:07:00', false],
            // 15th minute, of the second hour, every 15 days, in January, every Friday
            ['1 * * * 7', '2015-08-10 21:47:27', '2015-08-16 00:01:00', false],
            // Test with exact times
            ['47 21 * * *', strtotime('2015-08-10 21:47:30'), '2015-08-10 21:47:00', true],
            // Test Day of the week (issue #1)
            // According cron implementation, 0|7 = sunday, 1 => monday, etc
            ['* * * * 0', strtotime('2011-06-15 23:09:00'), '2011-06-19 00:00:00', false],
            ['* * * * 7', strtotime('2011-06-15 23:09:00'), '2011-06-19 00:00:00', false],
            ['* * * * 1', strtotime('2011-06-15 23:09:00'), '2011-06-20 00:00:00', false],
            // Should return the sunday date as 7 equals 0
            ['0 0 * * MON,SUN', strtotime('2011-06-15 23:09:00'), '2011-06-19 00:00:00', false],
            ['0 0 * * 1,7', strtotime('2011-06-15 23:09:00'), '2011-06-19 00:00:00', false],
            ['0 0 * * 0-4', strtotime('2011-06-15 23:09:00'), '2011-06-16 00:00:00', false],
            ['0 0 * * 7-4', strtotime('2011-06-15 23:09:00'), '2011-06-16 00:00:00', false],
            ['0 0 * * 4-7', strtotime('2011-06-15 23:09:00'), '2011-06-16 00:00:00', false],
            ['0 0 * * 7-3', strtotime('2011-06-15 23:09:00'), '2011-06-19 00:00:00', false],
            ['0 0 * * 3-7', strtotime('2011-06-15 23:09:00'), '2011-06-16 00:00:00', false],
            ['0 0 * * 3-7', strtotime('2011-06-18 23:09:00'), '2011-06-19 00:00:00', false],
            // Test lists of values and ranges (Abhoryo)
            ['0 0 * * 2-7', strtotime('2011-06-20 23:09:00'), '2011-06-21 00:00:00', false],
            ['0 0 * * 0,2-6', strtotime('2011-06-20 23:09:00'), '2011-06-21 00:00:00', false],
            ['0 0 * * 2-7', strtotime('2011-06-18 23:09:00'), '2011-06-19 00:00:00', false],
            ['0 0 * * 4-7', strtotime('2011-07-19 00:00:00'), '2011-07-21 00:00:00', false],
            // Test increments of ranges
            ['0-12/4 * * * *', strtotime('2011-06-20 12:04:00'), '2011-06-20 12:04:00', true],
            ['4-59/2 * * * *', strtotime('2011-06-20 12:04:00'), '2011-06-20 12:04:00', true],
            ['4-59/2 * * * *', strtotime('2011-06-20 12:06:00'), '2011-06-20 12:06:00', true],
            ['4-59/3 * * * *', strtotime('2011-06-20 12:06:00'), '2011-06-20 12:07:00', false],
            ['0 0 * * 0,2-6', strtotime('2011-06-20 23:09:00'), '2011-06-21 00:00:00', false],
            // Test Day of the Week and the Day of the Month (issue #1)
            ['0 0 1 1 0', strtotime('2011-06-15 23:09:00'), '2012-01-01 00:00:00', false],
            ['0 0 1 JAN 0', strtotime('2011-06-15 23:09:00'), '2012-01-01 00:00:00', false],
            ['0 0 1 * 0', strtotime('2011-06-15 23:09:00'), '2012-01-01 00:00:00', false],
            ['0 0 L * *', strtotime('2011-07-15 00:00:00'), '2011-07-31 00:00:00', false],
            // Test the W day of the week modifier for day of the month field
            ['0 0 2W * *', strtotime('2011-07-01 00:00:00'), '2011-07-01 00:00:00', true],
            ['0 0 1W * *', strtotime('2011-05-01 00:00:00'), '2011-05-02 00:00:00', false],
            ['0 0 1W * *', strtotime('2011-07-01 00:00:00'), '2011-07-01 00:00:00', true],
            ['0 0 3W * *', strtotime('2011-07-01 00:00:00'), '2011-07-04 00:00:00', false],
            ['0 0 16W * *', strtotime('2011-07-01 00:00:00'), '2011-07-15 00:00:00', false],
            ['0 0 28W * *', strtotime('2011-07-01 00:00:00'), '2011-07-28 00:00:00', false],
            ['0 0 30W * *', strtotime('2011-07-01 00:00:00'), '2011-07-29 00:00:00', false],
            ['0 0 31W * *', strtotime('2011-07-01 00:00:00'), '2011-07-29 00:00:00', false],
            // Test the year field
            ['* * * * * 2012', strtotime('2011-05-01 00:00:00'), '2012-01-01 00:00:00', false],
            // Test the last weekday of a month
            ['* * * * 5L', strtotime('2011-07-01 00:00:00'), '2011-07-29 00:00:00', false],
            ['* * * * 6L', strtotime('2011-07-01 00:00:00'), '2011-07-30 00:00:00', false],
            ['* * * * 7L', strtotime('2011-07-01 00:00:00'), '2011-07-31 00:00:00', false],
            ['* * * * 1L', strtotime('2011-07-24 00:00:00'), '2011-07-25 00:00:00', false],
            ['* * * * TUEL', strtotime('2011-07-24 00:00:00'), '2011-07-26 00:00:00', false],
            ['* * * 1 5L', strtotime('2011-12-25 00:00:00'), '2012-01-27 00:00:00', false],
            // Test the hash symbol for the nth weekday of a given month
            ['* * * * 5#2', strtotime('2011-07-01 00:00:00'), '2011-07-08 00:00:00', false],
            ['* * * * 5#1', strtotime('2011-07-01 00:00:00'), '2011-07-01 00:00:00', true],
            ['* * * * 3#4', strtotime('2011-07-01 00:00:00'), '2011-07-27 00:00:00', false],
            ['5/0 * * * *', time(), '2011-07-27 00:00:00', false],
            ['5/20 * * * *', '2018-08-13 00:24:00', '2011-07-27 00:00:00', false],                // issue #12
            ['5/20 * * * *', strtotime('2018-08-13 00:45:00'), '2011-07-27 00:00:00', true],      // issue #12
            ['5-11/4 * * * *', strtotime('2018-08-13 00:03:00'), '2011-07-27 00:00:00', false],

            // New from https://github.com/dragonmantank/cron-expression
            ['0 0 L * 0', strtotime('2011-06-15 23:09:00'), '2011-06-19 00:00:00', false],
            ['3-59/15 6-12 */15 1 2-5', strtotime('2017-01-08 00:00:00'), '2017-01-10 06:03:00', false],
            ['* * * * MON-FRI', strtotime('2017-01-08 00:00:00'), strtotime('2017-01-09 00:00:00'), false],
            ['* * * * TUE', strtotime('2017-01-08 00:00:00'), strtotime('2017-01-10 00:00:00'), false],
            ['0 1 15 JUL mon,Wed,FRi', strtotime('2019-11-14 00:00:00'), strtotime('2020-07-01 01:00:00'), false],
            ['0 1 15 jul mon,Wed,FRi', strtotime('2019-11-14 00:00:00'), strtotime('2020-07-01 01:00:00'), false],
            ['@weekly', strtotime('2019-11-14 00:00:00'), strtotime('2019-11-17 00:00:00'), false],
            ['@weekly', strtotime('2019-11-14 00:00:00'), strtotime('2019-11-17 00:00:00'), false],
            ['@weekly', strtotime('2019-11-14 00:00:00'), strtotime('2019-11-17 00:00:00'), false],
            ['0 12 * * ?', strtotime('2020-08-20 00:00:00'), strtotime('2020-08-20 12:00:00'), false],
            ['0 12 ? * *', strtotime('2020-08-20 00:00:00'), strtotime('2020-08-20 12:00:00'), false],
        ];
    }
}


src/Validator.php
<?php

declare(strict_types=1);

/*
 * This file is part of the PHP-CRON-EXPR package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\Cron;

/**
 * Cron segment validator.
 *
 * This class checks if a cron segment is valid.
 *
 * @author Jitendra Adhikari <jiten.adhikary@gmail.com>
 */
class Validator
{
    /**
     * Check if the value is in range of given offset.
     *
     * @param int    $value
     * @param string $offset
     *
     * @return bool
     */
    public function inRange(int $value, string $offset): bool
    {
        $parts = \explode('-', $offset);

        return $parts[0] <= $value && $value <= $parts[1];
    }

    /**
     * Check if the value is in step of given offset.
     *
     * @param int    $value
     * @param string $offset
     *
     * @return bool
     */
    public function inStep(int $value, string $offset): bool
    {
        $parts = \explode('/', $offset, 2);

        if (empty($parts[1])) {
            return false;
        }

        if (\strpos($offset, '*/') === 0 || \strpos($offset, '0/') === 0) {
            return $value % $parts[1] === 0;
        }

        $subparts = \explode('-', $parts[0], 2) + [1 => $value];

        return $this->inStepRange((int) $value, (int) $subparts[0], (int) $subparts[1], (int) $parts[1]);
    }

    /**
     * Check if the value falls between start and end when advanved by step.
     *
     * @param int $value
     * @param int $start
     * @param int $end
     * @param int $step
     *
     * @return bool
     */
    public function inStepRange(int $value, int $start, int $end, int $step): bool
    {
        if (($start + $step) > $end) {
            return false;
        }

        if ($start <= $value && $value <= $end) {
            return \in_array($value, \range($start, $end, $step));
        }

        return false;
    }

    /**
     * Check if month modifiers [L C W #] are satisfied.
     *
     * @internal
     *
     * @param string        $value
     * @param ReferenceTime $reference
     *
     * @return bool
     */
    public function isValidMonthDay(string $value, ReferenceTime $reference): bool
    {
        if ($value == 'L') {
            return $reference->monthDay() == $reference->numDays();
        }

        if ($pos = \strpos($value, 'W')) {
            $value = \substr($value, 0, $pos);
            $month = $this->zeroPad($reference->month());

            return $this->isClosestWeekDay((int) $value, $month, $reference);
        }

        $this->unexpectedValue(2, $value);
        // @codeCoverageIgnoreStart
    }

    // @codeCoverageIgnoreEnd

    protected function isClosestWeekDay(int $value, string $month, ReferenceTime $reference): bool
    {
        foreach ([0, -1, 1, -2, 2] as $i) {
            $incr = $value + $i;
            if ($incr < 1 || $incr > $reference->numDays()) {
                continue;
            }

            $incr  = $this->zeroPad($incr);
            $parts = \explode(' ', \date('N m j', \strtotime("{$reference->year()}-$month-$incr")));
            if ($parts[0] < 6 && $parts[1] == $month) {
                return $reference->monthDay() == $parts[2];
            }
        }

        // @codeCoverageIgnoreStart
        return false;
        // @codeCoverageIgnoreEnd
    }

    /**
     * Check if week modifiers [L C W #] are satisfied.
     *
     * @internal
     *
     * @param string        $value
     * @param ReferenceTime $reference
     *
     * @return bool
     */
    public function isValidWeekDay(string $value, ReferenceTime $reference): bool
    {
        $month = $this->zeroPad($reference->month());

        if (\strpos($value, 'L')) {
            return $this->isLastWeekDay($value, $month, $reference);
        }

        if (!\strpos($value, '#')) {
            $this->unexpectedValue(4, $value);
        }

        list($day, $nth) = \explode('#', \str_replace('7#', '0#', $value));

        if (!$this->isNthWeekDay((int) $day, (int) $nth) || $reference->weekDay() != $day) {
            return false;
        }

        return \intval($reference->day() / 7) == $nth - 1;
    }

    /**
     * Throws UnexpectedValueException.
     *
     * @param int    $pos
     * @param string $value
     *
     * @throws \UnexpectedValueException
     */
    public function unexpectedValue(int $pos, string $value)
    {
        throw new \UnexpectedValueException(
            \sprintf('Invalid offset value at segment #%d: %s', $pos, $value)
        );
    }

    protected function isLastWeekDay(string $value, string $month, ReferenceTime $reference): bool
    {
        $decr  = $reference->numDays();
        $value = \explode('L', \str_replace('7L', '0L', $value));

        for ($i = 0; $i < 7; $i++) {
            $decr -= $i;
            if (\date('w', \strtotime("{$reference->year()}-$month-$decr")) == $value[0]) {
                return $reference->monthDay() == $decr;
            }
        }

        return false;
    }

    protected function isNthWeekDay(int $day, int $nth): bool
    {
        return !($day < 0 || $day > 7 || $nth < 1 || $nth > 5);
    }

    protected function zeroPad($value): string
    {
        return \str_pad((string) $value, 2, '0', \STR_PAD_LEFT);
    }
}


src/Expression.php
<?php

declare(strict_types=1);

/*
 * This file is part of the PHP-CRON-EXPR package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\Cron;

/**
 * Cron Expression Parser.
 *
 * This class checks if a cron expression is due to run on given timestamp (or default now).
 * Acknowledgement: The initial idea came from {@link http://stackoverflow.com/a/5727346}.
 *
 * @author Jitendra Adhikari <jiten.adhikary@gmail.com>
 */
class Expression
{
    /** @var Expression */
    protected static $instance;

    /** @var SegmentChecker */
    protected $checker;

    /** @var Normalizer */
    protected $normalizer;

    public function __construct(SegmentChecker $checker = null, Normalizer $normalizer = null)
    {
        $this->checker    = $checker ?: new SegmentChecker;
        $this->normalizer = $normalizer ?: new Normalizer;

        if (null === static::$instance) {
            static::$instance = $this;
        }
    }

    public static function instance(): self
    {
        if (null === static::$instance) {
            static::$instance = new static;
        }

        return static::$instance;
    }

    /**
     * Parse cron expression to decide if it can be run on given time (or default now).
     *
     * @param string $expr The cron expression.
     * @param mixed  $time The timestamp to validate the cron expr against. Defaults to now.
     *
     * @return bool
     */
    public static function isDue(string $expr, $time = null): bool
    {
        return static::instance()->isCronDue($expr, $time);
    }

    /**
     * Filter only the jobs that are due.
     *
     * @param array $jobs Jobs with cron exprs. [job1 => cron-expr1, job2 => cron-expr2, ...]
     * @param mixed $time The timestamp to validate the cron expr against. Defaults to now.
     *
     * @return array Due job names: [job1name, ...];
     */
    public static function getDues(array $jobs, $time = null): array
    {
        return static::instance()->filter($jobs, $time);
    }

    /**
     * Instance call.
     *
     * Parse cron expression to decide if it can be run on given time (or default now).
     *
     * @param string $expr The cron expression.
     * @param mixed  $time The timestamp to validate the cron expr against. Defaults to now.
     *
     * @return bool
     */
    public function isCronDue(string $expr, $time = null): bool
    {
        $this->checker->setReference(new ReferenceTime($time));

        foreach (\explode(' ', $this->normalizer->normalizeExpr($expr)) as $pos => $segment) {
            if ($segment === '*' || $segment === '?') {
                continue;
            }

            if (!$this->checker->checkDue($segment, $pos)) {
                return false;
            }
        }

        return true;
    }

    /**
     * Filter only the jobs that are due.
     *
     * @param array $jobs Jobs with cron exprs. [job1 => cron-expr1, job2 => cron-expr2, ...]
     * @param mixed $time The timestamp to validate the cron expr against. Defaults to now.
     *
     * @return array Due job names: [job1name, ...];
     */
    public function filter(array $jobs, $time = null): array
    {
        $dues = $cache = [];

        foreach ($jobs as $name => $expr) {
            $expr = $this->normalizer->normalizeExpr($expr);

            if (!isset($cache[$expr])) {
                $cache[$expr] = $this->isCronDue($expr, $time);
            }

            if ($cache[$expr]) {
                $dues[] = $name;
            }
        }

        return $dues;
    }
}


src/Normalizer.php
<?php

declare(strict_types=1);

/*
 * This file is part of the PHP-CRON-EXPR package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\Cron;

class Normalizer
{
    const YEARLY      = '@yearly';
    const ANNUALLY    = '@annually';
    const MONTHLY     = '@monthly';
    const WEEKLY      = '@weekly';
    const DAILY       = '@daily';
    const HOURLY      = '@hourly';
    const ALWAYS      = '@always';
    const FIVE_MIN    = '@5minutes';
    const TEN_MIN     = '@10minutes';
    const FIFTEEN_MIN = '@15minutes';
    const THIRTY_MIN  = '@30minutes';

    protected static $expressions = [
        self::YEARLY      => '0 0 1 1 *',
        self::ANNUALLY    => '0 0 1 1 *',
        self::MONTHLY     => '0 0 1 * *',
        self::WEEKLY      => '0 0 * * 0',
        self::DAILY       => '0 0 * * *',
        self::HOURLY      => '0 * * * *',
        self::ALWAYS      => '* * * * *',
        self::FIVE_MIN    => '*/5 * * * *',
        self::TEN_MIN     => '*/10 * * * *',
        self::FIFTEEN_MIN => '*/15 * * * *',
        self::THIRTY_MIN  => '0,30 * * * *',
    ];

    protected static $literals = [
        'sun' => 0,
        'mon' => 1,
        'tue' => 2,
        'wed' => 3,
        'thu' => 4,
        'fri' => 5,
        'sat' => 6,

        'jan' => 1,
        'feb' => 2,
        'mar' => 3,
        'apr' => 4,
        'may' => 5,
        'jun' => 6,
        'jul' => 7,
        'aug' => 8,
        'sep' => 9,
        'oct' => 10,
        'nov' => 11,
        'dec' => 12,
    ];

    public function normalizeExpr(string $expr): string
    {
        $expr = \trim($expr);

        if (isset(static::$expressions[$exp = \strtolower($expr)])) {
            return static::$expressions[$exp];
        }

        $expr  = \preg_replace('~\s+~', ' ', $expr);
        $count = \substr_count($expr, ' ');

        if ($count < 4 || $count > 5) {
            throw new \UnexpectedValueException(
                'Cron $expr should have 5 or 6 segments delimited by space'
            );
        }

        return \str_ireplace(\array_keys(static::$literals), \array_values(static::$literals), $expr);
    }
}
